pkg/cortex/modules.go

```diff
@@ -652,6 +654,64 @@ func (t *Cortex) initRulerStorage() (serv services.Service, err error) {
 	return
 }
 
+func createActiveQueryTracker(cfg querier.Config, logger log.Logger) promql.QueryTracker {
+	dir := cfg.ActiveQueryTrackerDir
+
+	if dir != "" {
+		return promql.NewActiveQueryTracker(dir, cfg.MaxConcurrent, logger)
+	}
+
+	return nil
+}
+
+func (t *Cortex) initRulerExternal() (serv services.Service, err error) {
+	if t.RulerStorage == nil {
+		level.Info(util_log.Logger).Log("msg", "RulerStorage is nil.  Not starting the ruler.")
+		return nil, nil
+	}
+
+	engine := promql.NewEngine(promql.EngineOpts{
+		Logger:             util_log.Logger,
+		Reg:                prometheus.DefaultRegisterer,
+		ActiveQueryTracker: createActiveQueryTracker(t.Cfg.Querier, util_log.Logger),
+		MaxSamples:         t.Cfg.Querier.MaxSamples,
+		Timeout:            t.Cfg.Querier.Timeout,
+		LookbackDelta:      t.Cfg.Querier.LookbackDelta,
+		EnablePerStepStats: t.Cfg.Querier.EnablePerStepStats,
+		EnableAtModifier:   t.Cfg.Querier.AtModifierEnabled,
+		NoStepSubqueryIntervalFn: func(int64) int64 {
+			return t.Cfg.Querier.DefaultEvaluationInterval.Milliseconds()
+		},
+	})
+	managerFactory := ruler.DefaultTenantManagerFactory(t.Cfg.Ruler, t.Cfg.ExternalPusher, t.Cfg.ExternalQueryable, engine, t.Overrides, prometheus.DefaultRegisterer)
```

Comment:
Before we pass the ExternalPusher and ExternalQueryable, I think we should validate that those are not nulls.